### ISSUE TITLE
error_messages should allow callable AS WELL AS strings.

### ISSUE THREAD

**Issue by ShayanArmanPercolate** [marshmallow-code/marshmallow#334]
Currently, if a field is required, we can state in error_messages what error message to raise if this field is not supplied. However, this error raised is a marshmallow ValidationError. What if we wish to raise a custom exception with a custom error when this field is missing.
- Problem 1
  Currently error_messages does not allow callables.
- Problem 2
  If required is set to True (required=True), validators do not get called for that field if the field is missing. Instead, this defaults to the supplied error_messages dict or the default dict.

Does anyone else think that this would be a useful feature?
@sloria @franciscod

**Comment by ShayanArmanPercolate** [comment]
@sloria This would be a very interesting feature. Right now, I don't think there's any way to throw a custom exception if for example, a field expects a List:

some_list = fields.List(fields.Str(), ..., error_messages=dict(type=callable_raise_serialization_exception())

Right now, the default is to throw a validation error.

**Comment by sloria** [comment]
Custom errors should be raised in validators. Indeed, "required" validation is a special case. Because of this, I wonder if it makes sense to allow `required` to be a callable.

``` python
req_field = fields.Field(required=lambda: raise MyError('required!'))
```

**Comment by ShayanArmanPercolate** [comment]
Yeah, in the case that you brought up, validators won't be invoked. Therefore I think that'll be useful to have. Or even have a callable return the error message instead of putting the message in there.

# Hat tip to django-rest-framework.

```
def fail(self, key, **kwargs):
    """A helper method that simply raises a `ValidationError`.
    """
    try:
        error = self.error_messages[key]
        msg = error if not callable(error) else error()
    except KeyError:
        class_name = self.__class__.__name__
        msg = MISSING_ERROR_MESSAGE.format(class_name=class_name, key=key)
        raise AssertionError(msg)
    if isinstance(msg, basestring):
        msg = msg.format(**kwargs)
    raise ValidationError(msg)
```

**Comment by sloria** [comment]
I think that's a fine approach. I wonder if it would be useful for the callable to receive the input value.

``` python
my_field = fields.Int(error_messages=dict(invalid=lambda v: '{v} is required'.format(v=v))
```

**Comment by ShayanArmanPercolate** [comment]
Yes I think that would be extremely useful. However, is there a way currently to access the value from a field declaration?

**Comment by sloria** [comment]
I suppose you could pass the value to `Field.fail`.

**Comment by repole** [comment]
It'd be beneficial for the sake of potential i18n for the callable to receive the field context, or perhaps the actual field itself.

``` python
def fail(self, key, **kwargs):
    """A helper method that simply raises a `ValidationError`.
    """
    try:
        error = self.error_messages[key]
        msg = error if not callable(error) else error(self.context, **kwargs)
    except KeyError:
        class_name = self.__class__.__name__
        msg = MISSING_ERROR_MESSAGE.format(class_name=class_name, key=key)
        raise AssertionError(msg)
    if isinstance(msg, basestring):
        msg = msg.format(**kwargs)
    raise ValidationError(msg)
```

With this being the case, i18n could be supported using this type of pattern:

``` python
def dummy_gettext(string, **variables):
    """Simple gettext stand in for when none is provided."""
    return string % variables

def format_error(context, **kwargs):
    self.context.get("gettext", dummy_gettext)("Invalid input format: %(input)s", **kwargs) 

my_field = fields.Int(error_messages=dict(format=format_error))
```

I'm not sure if that's the most graceful solution, but it would open up just about any custom error possibilities.

**Comment by justanr** [comment]
Does it make potential sense to make required a Validator? Backwards compatibility could be maintained by issuing a deprecation warning and adding `Required` to the validator list if the keyword is present.

**Comment by sloria** [comment]
> Does it make potential sense to make required a Validator?

@justanr This was also suggested in #78 . The rationale against making required a validator is [here](https://github.com/marshmallow-code/marshmallow/issues/78#issuecomment-65887190):

> The validation for required is handled separately because it must precede execution of the validators callables. If a required value is not passed, validation should fail fast (as there is no value to pass to the rest of the validators).

**Comment by sloria** [comment]
> It'd be beneficial for the sake of potential i18n for the callable to receive the field context, or perhaps the actual field itself.

@repole I think passing the field instance makes sense. So then we have the signature `lambda value, field: ...`

**Comment by sloria** [comment]
PRs for this welcome!

**Comment by deckar01** [comment]
I'm not sure I understand the use case for this. Customizing the required validation message in the error output? It seems like encapsulating the message strings into an i18n module would satisfy this use case and fix the same issue for the other hard coded message strings.

**Comment by lafrech** [comment]
Not sure how relevant this is, but for example, [umongo](https://github.com/Scille/umongo) (marshmallow-based Mongo ODM) does i18n in validation messages.

**Comment by sloria** [comment]
Closing this for now due to lack of interest/prioritization. Feel free to comment if you feel strongly about this

**Comment by repole** [comment]
Feel strongly enough that I started to go down the road of implementing this. With this being at the `Field` level, the only way to accomplish i18n without this type of change is to monkeypatch. If this is doable without breaking any existing code, I don't see a reason not to pursue it.

In my attempts to implement this, I did run into a few issues, worth noting:

1. There are 2 locations where Marshmallow is assuming error_messages[key] is a string (namely in the Email and URL fields). To allow callables, those locations had to be updated. Seeing as these are the only places generating an `error_message` without calling `make_error` in fields (and in some way that feels like an anti-pattern to begin with), this doesn't seem like a huge price to pay.

2. From a mypy type checking stand point, I have no idea what I'm doing. Having the `error_messages` dict have values that can be so inherently different (callable vs str) is not something that's simple to negotiate with the mypy checker (at least for me). Some guidance here would be much appreciated.

To elaborate on point 2, the errors I'm getting back from mypy are:
```
src\marshmallow\fields.py:250: error: "str" not callable
src\marshmallow\fields.py:251: error: Incompatible return value type (got "Union[str, Callable[..., Any]]", expected "str")
```

The lines in question are in a method that promises a `str` response and look like:
```
if callable(self.error_messages[key]):
    return self.error_messages[key](self, key, **kwargs)
return self.error_messages[key]
```

Have tried different typing annotations on `error_messages` as it's passed into Field, but I'm struggling to wrap my head around how to do this the right way.
